docs/hello_nf-core/03_add_module.md

```diff
@@ -1,63 +1,893 @@
+
+    // Generate ASCII art with cowpy (https://github.com/jeffbuttars/cowpy)
+    process cowpy {
+
+        publishDir 'results', mode: 'copy'
+
+        container 'community.wave.seqera.io/library/cowpy:1.1.5--3db457ae1977a273'
+        conda 'conda-forge::cowpy==1.1.5'
+
+        input:
+            path input_file
+            val character
+
+        output:
+            path "cowpy-${input_file}"
+
+        script:
+        """
+        cat $input_file | cowpy -c "$character" > cowpy-${input_file}
+        """
+    }
+    ```
+
+Key changes:
+
+1. **Input**: Changed from `path input_file` to `tuple val(meta), path(input_file)` to accept metadata
+2. **Output**: Changed to emit a tuple with metadata: `tuple val(meta), path("cowpy-${input_file}"), emit: cowpy_output`
+3. **Named emit**: Added `emit: cowpy_output` to give the output channel a descriptive name
+
+Now update the workflow to pass the tuple directly instead of extracting the file. Open [core-hello/workflows/hello.nf](core-hello/workflows/hello.nf):
+
+=== "After"
 
-TODO: instructions
+    ```groovy title="core-hello/workflows/hello.nf" linenums="39" hl_lines="2"
+        // generate ASCII art of the greetings with cowpy
+        cowpy(CAT_CAT.out.file_out, params.character)
+    ```
+
+=== "Before"
+
+    ```groovy title="core-hello/workflows/hello.nf" linenums="39" hl_lines="2-4"
+        // generate ASCII art of the greetings with cowpy
+        // Extract the file from the tuple since cowpy doesn't use metadata yet
+        ch_for_cowpy = CAT_CAT.out.file_out.map{ meta, file -> file }
+        cowpy(ch_for_cowpy, params.character)
+    ```
+
+Also update the emit block to use the named emit:
+
+=== "After"
+
+    ```groovy title="core-hello/workflows/hello.nf" linenums="58" hl_lines="2"
+        emit:
+        cowpy_hellos   = cowpy.out.cowpy_output
+        versions       = ch_versions                 // channel: [ path(versions.yml) ]
+    ```
+
+=== "Before"
+
+    ```groovy title="core-hello/workflows/hello.nf" linenums="58" hl_lines="2"
+        emit:
+        cowpy_hellos   = cowpy.out
+        versions       = ch_versions                 // channel: [ path(versions.yml) ]
+    ```
+
+Test the workflow to ensure metadata flows through correctly:
+
+```bash
+nextflow run . --outdir core-hello-results -profile test,docker --validate_params false
+```
+
+The pipeline should run successfully with metadata now flowing from `CAT_CAT` through `cowpy`.
+
+### 2.2. Simplify the interface with ext.args
+
+Now let's address another nf-core pattern: simplifying module interfaces by using `ext.args` for optional command-line arguments.
+
+Currently, our `cowpy` module requires the `character` parameter to be passed as a separate input. While this works, nf-core modules follow a convention of keeping interfaces minimal - only essential inputs (metadata and files) should be declared. Optional tool arguments are instead passed via configuration.
+
+#### Understanding ext.args
```

Comment:
OMG this was so helpful, I used to be confused about this. I have some thoughts about the ordering/breakdown of how this is presented (like above I would prefer to go a bit more step by step) but it does feel like the core material is all there.